BPMN comparison examples (#252)
serverlessworkflow/specification: comparisons/comparison-bpmn.md

# Comparisons - BPMN2

> The [Business Process Model and Notation (BPMN)](https://www.omg.org/spec/BPMN/2.0/PDF) defines a flowchart-based
> DSL for workflows. It is maintained by the [Object Management Group (OMG)](https://www.omg.org/). 
> The latest version of BPMN is [2.0.2](https://www.omg.org/spec/BPMN/2.0.2/), published in 2014.

Reviewer: ```suggestion
The [Business Process Model and Notation (BPMN)](https://www.omg.org/spec/BPMN/2.0/PDF) defines a flowchart-based
DSL for workflows. It is maintained by the [Object Management Group (OMG)](https://www.omg.org/). 
The latest BPMN version is [2.0.2](https://www.omg.org/spec/BPMN/2.0.2/), published in 2014.
```

> BPMN2 defines a graphical notation to specify workflows. This notation can then be shared between tooling and organizations.
> The graphical notation is translated into XML which then can be used for execution on runtime engines.

Reviewer: ```suggestion
BPMN2 defines a graphical notation to specify workflows. This notation can then be shared between tooling and organizations.
The graphical notation is translated into XML, which can then be used to execute runtime engines.
```

Author: "to execute runtime engines" sounds weird imo. any other suggestions ?

> For this comparison we will compare the Serverless Workflow language with the graphical representation of BPMN2,
> and not its underlying XML representation. The BPMN2 XML is very difficult to understand, quite large for even the smallest 
> workflows, and often times not portable between runtimes. 
> It makes more sense to use its portable graphical notation for comparisons.

Reviewer: ```suggestion
For this comparison, we will compare the Serverless Workflow language with the graphical representation of BPMN2,
and not its underlying XML representation. The BPMN2 XML is very difficult to understand, quite large for even the smallest workflows, and often not portable between runtimes. 
It makes more sense to use its portable graphical notation for comparisons.
```

> Serverless Workflow is a declarative workflow language, represented with JSON or YAML. It currently does not 
> define a graphical notation however can be graphically represented using different flowcharting techniques such as 
> UML activity diagrams. The [Serverless Workflow Java SDK](https://github.com/serverlessworkflow/sdk-java#building-workflow-diagram) 
> as well as its [VSCode Extension](https://github.com/serverlessworkflow/vscode-extension) provide means to generate SVG diagrams based on the workflow
> JSON/YAML. 

Reviewer: ```suggestion
Serverless Workflow is a declarative workflow language, represented with JSON or YAML. It currently does not define a graphical notation. However, it can be graphically represented using different flowcharting techniques such as 
UML activity diagrams. The [Serverless Workflow Java SDK](https://github.com/serverlessworkflow/sdk-java#building-workflow-diagram) as well as its [VSCode Extension](https://github.com/serverlessworkflow/vscode-extension) provide means to generate SVG diagrams based on the workflow JSON/YAML. 
```

## Note when reading provided examples

> The BPMN2 graphical notation does not provide details about things like data inputs/outputs, mapping and transformation. 
> BPMN2 does provide graphical representation for this such as Data Objects, however most of the examples
> available do not use them. Execution semantics such as task and event properties are also not visual.
> For this reason the event, function, retry, and data mapping 
> defined in the associated Serverless Workflow YAML are assumed. 

Reviewer: ```suggestion
The BPMN2 graphical notation does not provide details about data inputs/outputs, mapping, and transformation. 
BPMN2 does provide graphical representation for things such as Data Objects. However, most of the examples
available do not use them. Execution semantics such as task and event properties are also not visual.
For this reason, the event, function, retry, and data mapping defined in the associated Serverless Workflow YAML are assumed. 
```

  interval: PT7D
  interrupt: true
  runBefore: Handle timeout
states:
- name: Purchase Parts
  start: true
  type: operation
  actions:
  - functionRef: purchasePartsFunction
  transition: Unpack Parts
- name: Unpack Parts
  type: operation
  actions:
  - functionRef: unpackPartsFunction
  end: true
- name: Handle timeout
  type: operation
  actions:
  - functionRef: handleTimeoutFunction
functions:
- name: purchasePartsFunction
  operation: file://myservice.json#purchase
- name: unpackPartsFunction
  operation: file://myservice.json#unpack
- name: handleTimeoutFunction
  operation: file://myservice.json#handle
```

</td>
</tr>
</table>

### Multiple Instance Subprocess

<table>
<tr>
    <th>BPMN2 Diagram</th>
    <th>Serverless Workflow</th>
</tr>
<tr>
<td valign="top">
<p align="center">
<img src="../media/comparisons/bpmn/multiinstance-subprocess.png" width="500px" alt="BPMN2 Multi-Instance Subprocess Workflow"/>
</p>
</td>
<td valign="top">

```yaml
id: subflowloop
name: SubFlow Repeat workflow
version: '1.0'
states:
- name: SubflowRepeat
  type: subflow
  start: true
  workflowId: dosomethingandwaitformessage
  repeat:
    max: 10
  end: true
```

</td>
</tr>
</table>

> * Note: We did not include the `dosomethingandwaitformessage` workflow in this example which would just include
> a starting "operation" state transitioning to an "event" state that waits for the needed event.

Reviewer: ```suggestion
* Note: We did not include the `dosomethingandwaitformessage` workflow in this example, which would just include
a starting "operation" state transitioning to an "event" state that waits for the needed event.
```
